lib/binding/kvdb_interface.c

```diff
@@ -95,24 +99,38 @@ hse_lowmem_adjust(unsigned long *memgb)
 hse_err_t
 hse_init(const char *const config, const size_t paramc, const char *const *const paramv)
 {
-    merr_t         err;
-    ulong          memgb;
+    int rc;
+    ulong memgb;
+    merr_t err = 0;
     struct config *conf;
-
-    if (hse_initialized)
-        return 0;
+    bool lock_taken = false;
 
     if (HSE_UNLIKELY(paramc > 0 && !paramv))
         return merr(EINVAL);
 
-    hse_progname = program_invocation_name ?: __func__;
+    rc = sem_init(&hse_lock, 0, 1);
+    if (rc == -1)
+        return merr(errno);
+
+    rc = sem_wait(&hse_lock);
+    if (rc == -1) {
+        err = merr(errno);
+        goto out;
+    }
+
+    lock_taken = true;
+
+    if (hse_initialized)
+        goto out;
```

Comment:
abort() ?

Reply:
This is to protect against someone calling `hse_init()` twice. I don't think abort() is the right answer here. This is also what was there previously.

Reply:
Right, but at this point you reinitialized the sema, so something funky is going on..

Reply:
> Initializing a semaphore that has already been initialized results in undefined behavior.

@alexttx what are your thoughts on just aborting here?

Reply:
abort seems wrong.  using a dynamically initialized synchronization object add challenges.  what if an API is called before hse_init is called and tries to use the uninitialized semaphore?  admittedly, that's a misuse of the API, but that's what this PR is trying to address.

Reply:
Yea, the semaphore not having a static initializer is very annoying.

This PR is trying to address thread safety on our part. This is really just a straight hack until we decide to implement thread safety better.

Reply:
I guess just use an r/w lock then...  You could initialize the sema in a constructor, but we don't wanna go there..

Reply:
You could use a mutex, but it feels dirty...  I guess it'd be ok, there's not interrupting it though...